ws-client/src/client.rs

```diff
@@ -35,48 +35,65 @@ use jsonrpsee_types::{
 	error::Error,
 	jsonrpc::{self, JsonValue, SubscriptionId},
 };
-use std::convert::TryInto;
 use std::time::Duration;
+use std::{borrow::Cow, convert::TryInto};
 use std::{io, marker::PhantomData};
 
 /// Client that can be cloned.
 ///
-/// > **Note**: This struct is designed to be easy to use, but it works by maintaining a background
-/// >           task running in parallel. If this is not desirable, you are encouraged to use the
-/// >           [`RawClient`] struct instead.
+/// > **Note**: This struct is designed to be easy to use, but it works by maintaining a background task running in parallel.
 #[derive(Clone)]
 pub struct WsClient {
 	/// Channel to send requests to the background task.
 	to_back: mpsc::Sender<FrontToBack>,
-	/// Config.
-	config: WsConfig,
+	/// Request timeout
+	request_timeout: Option<Duration>,
 }
 
-#[derive(Copy, Clone, Debug)]
 /// Configuration.
-pub struct WsConfig {
-	/// Backend channel for serving requests and notifications.
-	pub request_channel_capacity: usize,
-	/// Backend channel for each unique subscription.
-	pub subscription_channel_capacity: usize,
+#[derive(Clone, Debug)]
+pub struct WsConfig<'a> {
+	/// URL to connect to.
+	pub url: &'a str,
 	/// Max request body size
 	pub max_request_body_size: usize,
 	/// Request timeout
 	pub request_timeout: Option<Duration>,
+	/// Connection timeout
+	pub connection_timeout: Duration,
+	/// `Origin` header to pass during the HTTP handshake. If `None`, no
+	/// `Origin` header is passed.
+	pub origin: Option<Cow<'a, str>>,
+	/// Url to send during the HTTP handshake.
+	pub handshake_url: Cow<'a, str>,
+	/// Max concurrent request capacity.
+	// TODO(niklasad1): clarify that the size is the size of initial mpsc::channel
+	// which may be increased if the sender is cloned.
+	pub max_concurrent_requests_capacity: usize,
+	/// Max concurrent capacity for each subscription.
+	/// If this limit is exceeded the subscription will be dropped.
+	// TODO(niklasad1): clarify that the size is the size of initial mpsc::channel
+	// which may be increased if the sender is cloned.
+	pub max_subscription_capacity: usize,
 }
 
-impl Default for WsConfig {
-	fn default() -> Self {
+impl<'a> WsConfig<'a> {
+	/// Default `WebSocket` configuration with a specified URL to connect to.
+	pub fn with_url(url: &'a str) -> Self {
 		Self {
-			request_channel_capacity: 100,
-			subscription_channel_capacity: 4,
+			url,
 			max_request_body_size: 10 * 1024 * 1024,
 			request_timeout: None,
+			connection_timeout: Duration::from_secs(10),
+			origin: None,
+			handshake_url: From::from("/"),
+			max_concurrent_requests_capacity: 100,
```

Comment:
maybe decrease this to something like 10?

Reply:
10 sounds pretty low. What's the `size_of` the messages we are passing?

Reply:
Yeah, your hunch is correct. 
The message size is 96 bytes.

So 10 * 96  is quite reasonable, but we could to `10 * + std::mem::size_of::<FrontToBack>()` to future proof it.

Reply:
Wait, no, that doesn't make any sense, the channel capacity is in messages, not in bytes :upside_down_face: 

I'm asking what the size of the message is to make sure the buffer isn't going to cause memory issues. If we set the default to be something like 256, then (with current shape of the message) that buffer will allocate some 24kb, which should be fine.

Reply:
lol, I got confused thanks.